### PR TITLE
Lab 1 - Fibonacci and Negafibonacci

### PR DESCRIPTION
**Pull request by kasunfdo**
Fixes #24 

#### Changes proposed in this PR:
* Implement Fibonacci and Negafibonacci series
* Add unit tests

### CODE REVIEW

File: 01_fib/kasunfdo/main.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+func abs(x int) int {
+	if x < 0 {
+		return -x
+	}
+	return x
+}
+
+func fib(n int) {
+	if n == 0 {
+		_, _ = fmt.Fprintln(out, 0)
```
Comment (marcelocantos) [inline 01_fib/kasunfdo/main.go]: No need for assignment.

File: 01_fib/kasunfdo/main.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+		return
+	}
+
+	absVal := abs(n)
+	sign := n / absVal
+	n1 := 1
+	n2 := sign
```
Comment (marcelocantos) [inline 01_fib/kasunfdo/main.go]: `n1, n2 := 1, sign` (to match the assignment inside the loop)

File: 01_fib/kasunfdo/main.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+	for j := 0; j < absVal; j++ {
+		_, _ = fmt.Fprintln(out, n1)
```
Comment (marcelocantos) [inline 01_fib/kasunfdo/main.go]: No need for assignment.

File: 01_fib/kasunfdo/main_test.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+-3
+5
+-8
+13
+`)
+	actual := strconv.Quote(buf.String())
+
+	fmt.Println(expected)
+	fmt.Println(actual)
+	fmt.Println(expected != actual)
```
Comment (marcelocantos) [inline 01_fib/kasunfdo/main_test.go]: Don't print in tests unless there's a problem.

File: 01_fib/kasunfdo/main_test.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	if expected != actual {
+		t.Errorf("Unexpected output in main()")
```
Comment (marcelocantos) [inline 01_fib/kasunfdo/main_test.go]: Print full context here. Something like:
```go
t.Errorf("Unexpected output in main()\nexpected: %q\nactual: %q", expected, actual)
```

**Review by n0npax — DISMISSED** [review]
LGTM

**Comment by kasunfdo** [comment]
Rebased with latest master branch. Reapproval requires.

**Review by juliaogris — CHANGES_REQUESTED** [review]
This might be a bit nit-picky - over all this PR looks pretty perfect to me, but seeing as you are the first one to submit it: Most of these tests here do exactly the same thing, so use Table Driven tests instead (maps as data structure and `t.Run(...)` as described in recent [Dave Cheney post](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) at the bottom: "Introducing sub tests").

**Comment by kasunfdo** [comment]
Thank you @juliaogris for guidance. I updated the PR.

**Review by juliaogris — APPROVED** [review]
LGTM!

You have done a lot more than was asked of you - respect!

Nonetheless, for future labs your table driven tests would look even nicer if you used something similar to this [sample code](https://github.com/anz-bank/go-slides/blob/master/samples/testing/basic/cases_test.go#L14) - see how it uses `t.Run` and a `map` for the actual test cases? This will print test cases nicer and make it easier to debug in case of failure (also described in the bottom section of the [Dave Cheney post](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests)).

**Comment by kasunfdo** [comment]
Noted. Thanks

**Pull request merged**